File: mycelium-py/src/mycelium/wiki.py
"""
mycelium.wiki — HTTP gateway client for the Mycelium group wiki.

Wraps the ``/gateway/wiki/*`` endpoints exposed by a Rust node running the
``mycelium-wiki`` companion crate with the ``gateway`` feature.

The wiki is the group's durable, curated knowledge canon — the long-term-memory
sibling of the blackboard's working memory. ``read``/``query`` are served
directly from the store (any node, in parallel); ``propose`` enqueues an edit
that the group's elected **curator** applies (single writer of record). It
composes with an external metrics store (Postgres) and RAG by a shared id
namespace — it is the *authoritative, maintained-meaning* layer, not a
similarity index.

Example::

    import asyncio
    from mycelium.wiki import Wiki

    async def main():
        wiki = Wiki("127.0.0.1", 7946, group="council")
        await wiki.propose(page="decisions/elm-street",
                           heading="Resolution 2026-14",
                           body="protected bike lane approved",
                           attributes={"topic": "transport"})
        page = await wiki.read("decisions/elm-street")     # served from the store
        hits = await wiki.query(equals={"topic": "transport"})

    asyncio.run(main())
"""

from __future__ import annotations

from typing import Optional

import httpx

from ._pool import ClientPool, PoolOwner


class Wiki(PoolOwner):
    """Async client for one group's wiki via a node's HTTP gateway."""

    def __init__(self, host: str, port: int, group: str = "wiki", *, token: Optional[str] = None):
        self._base_url = f"http://{host}:{port}"
        self._group = group
        self._pool = ClientPool(self._base_url, token=token)

    async def read(self, page: str) -> Optional[dict]:
        """Read a page (manifest joined with its live sections, in render order), or ``None`` if the
        page has no manifest. Served directly from the store."""
        async with self._pool.asy() as c:
            r = await c.post("/gateway/wiki/read", json={"group": self._group, "page": page})
        r.raise_for_status()
        return r.json()["page"]

    async def query(self, equals: Optional[dict[str, str]] = None) -> list[dict]:
        """Query sections by attribute (all-of equality — a structured filter, not similarity search).
        Returns ``[{page, id, heading, attributes}, …]``."""
        async with self._pool.asy() as c:
            r = await c.post("/gateway/wiki/query", json={"group": self._group, "equals": equals or {}})
        r.raise_for_status()
        return r.json()["hits"]

    async def propose(
        self,
        page: str,
        body: str,
        heading: str = "",
        section: Optional[str] = None,
        attributes: Optional[dict[str, str]] = None,
    ) -> dict:
        """Propose an edit; the curator applies it (single writer of record). Omit ``section`` to mint a
        new one, or pass an existing id to edit it. Returns ``{"proposal": key, "section": id}``."""
        payload = {
            "group": self._group,
            "page": page,
            "heading": heading,
            "body": body,
            "attributes": attributes or {},
        }
        if section is not None:
            payload["section"] = section
        async with self._pool.asy() as c:
            r = await c.post("/gateway/wiki/propose", json=payload)
        r.raise_for_status()
        return r.json()

    async def ingest(self, reference: str, timeout_secs: int = 60) -> dict:
        """Submit a staged batch's claim-check *reference* for bulk ingest (council-substrate
        Phase 4). The payload never rides the request — the batch is staged in the curator's
        ``BatchSource`` (S3 in the council deployment); the curator fetches, applies the whole
        batch atomically through its write gate, publishes, and returns the summary:
        ``{"summary": {"applied": n, "refused": n, "findings": [...]}}``. Sizing contract:
        a batch = one meeting."""
        payload = {"group": self._group, "reference": reference, "timeout_secs": timeout_secs}
        async with self._pool.asy(timeout=timeout_secs + 15.0) as c:
            r = await c.post("/gateway/wiki/ingest", json=payload)
        r.raise_for_status()
        return r.json()

File: mycelium-ts/src/wiki.ts
import { authHeaders, resolveToken, type AuthOptions } from "./auth";
/**
 * mycelium/wiki — TypeScript client for the Mycelium group wiki.
 *
 * Wraps the `/gateway/wiki/*` endpoints exposed by a Rust node running the
 * `mycelium-wiki` companion crate with the `gateway` feature.
 *
 * The wiki is the group's durable, curated knowledge canon — the long-term-memory
 * sibling of the blackboard's working memory. `read`/`query` are served directly
 * from the store (any node); `propose` enqueues an edit that the group's elected
 * curator applies (single writer of record). It composes with an external metrics
 * store (Postgres) and RAG by a shared id namespace — it is the *authoritative,
 * maintained-meaning* layer, not a similarity index.
 *
 * @example
 * ```ts
 * import { Wiki } from "mycelium-ts";
 * const wiki = new Wiki("127.0.0.1", 7946, "council");
 * const { proposal } = await wiki.propose({ page: "decisions/elm-street", heading: "Resolution 2026-14", body: "bike lane approved" });
 * const page = await wiki.read("decisions/elm-street");
 * const hits = await wiki.query({ topic: "transport" });
 * ```
 */

/** A section as read: stable opaque id, editable heading + body, and join-key/scope attributes. */
export interface Section {
  id: string;
  heading: string;
  body: string;
  attributes: Record<string, string>;
}

/** A page as read: the manifest joined with its live sections, in render order. */
export interface Page {
  path: string;
  attributes: Record<string, string>;
  sections: Section[];
}

/** A lightweight query hit — which section on which page matched, without the full-page join. */
export interface SectionRef {
  page: string;
  id: string;
  heading: string;
  attributes: Record<string, string>;
}

export interface ProposeArgs {
  page: string;
  /** Omit to mint a new section; provide an existing id to edit it. */
  section?: string;
  heading?: string;
  body: string;
  attributes?: Record<string, string>;
}

/** Async client for one group's wiki via a node's HTTP gateway. */
export class Wiki {
  private readonly baseUrl: string;
  private readonly auth: Record<string, string>;
  constructor(
    host: string,
    port: number,
    private readonly group: string = "wiki",
    opts: AuthOptions = {},
  ) {
    this.baseUrl = `http://${host}:${port}`;
    this.auth = authHeaders(resolveToken(opts.token));
  }

  /** Read a page (manifest + live sections), or `null` if it has no manifest. Served directly from the store. */
  async read(page: string): Promise<Page | null> {
    const r = await fetch(`${this.baseUrl}/gateway/wiki/read`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...this.auth },
      body: JSON.stringify({ group: this.group, page }),
    });
    if (!r.ok) throw new Error(`wiki read failed: ${r.status}`);
    return ((await r.json()) as { page: Page | null }).page;
  }

  /** Query sections by attribute (all-of equality — structured filter, not similarity search). */
  async query(equals: Record<string, string> = {}): Promise<SectionRef[]> {
    const r = await fetch(`${this.baseUrl}/gateway/wiki/query`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...this.auth },
      body: JSON.stringify({ group: this.group, equals }),
    });
    if (!r.ok) throw new Error(`wiki query failed: ${r.status}`);
    return ((await r.json()) as { hits: SectionRef[] }).hits;
  }

  /** Propose an edit; the curator applies it. Returns the proposal key + the (minted or given) section id. */
  async propose(args: ProposeArgs): Promise<{ proposal: string; section: string }> {
    const r = await fetch(`${this.baseUrl}/gateway/wiki/propose`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...this.auth },
      body: JSON.stringify({ group: this.group, ...args }),
    });
    if (!r.ok) throw new Error(`wiki propose failed: ${r.status}`);
    return (await r.json()) as { proposal: string; section: string };
  }

  /** Submit a staged batch.s claim-check reference for bulk ingest (council-substrate Phase 4).
   * The payload never rides the request — the curator fetches from its own BatchSource, applies
   * the whole batch atomically through its write gate, and returns the summary. Sizing contract:
   * a batch = one meeting. */
  async ingest(reference: string, timeoutSecs = 60): Promise<{ summary: { applied: number; refused: number; findings: string[] } }> {
    const r = await fetch(`${this.baseUrl}/gateway/wiki/ingest`, {
      method: "POST",
      headers: { "Content-Type": "application/json", ...this.auth },
      body: JSON.stringify({ group: this.group, reference, timeout_secs: timeoutSecs }),
    });
    if (!r.ok) throw new Error(`wiki ingest failed: ${r.status}`);
    return (await r.json()) as { summary: { applied: number; refused: number; findings: string[] } };
  }
}
